File: 4.0.0/com.android.tools/sdk-common/com/android/ide/common/gradle/model/level2/IdeJavaLibrary.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model.level2;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.builder.model.level2.Library;
import java.io.File;
import java.io.Serializable;
import java.util.Collection;
import java.util.Objects;

/** Creates a deep copy of {@link Library} of type LIBRARY_JAVA. */
public final class IdeJavaLibrary implements Library, Serializable {
    // Increase the value when adding/removing fields or when changing the serialization/deserialization mechanism.
    private static final long serialVersionUID = 3L;

    @NonNull private final String myArtifactAddress;
    @NonNull private final File myArtifactFile;
    private final int myType;
    private final int myHashCode;

    // Used for serialization by the IDE.
    @SuppressWarnings("unused")
    IdeJavaLibrary() {
        myArtifactAddress = "";
        //noinspection ConstantConditions
        myArtifactFile = null;
        myType = 0;

        myHashCode = 0;
    }

    IdeJavaLibrary(@NonNull String artifactAddress, @NonNull File artifactFile) {
        myType = LIBRARY_JAVA;
        myArtifactAddress = artifactAddress;
        myArtifactFile = artifactFile;
        myHashCode = calculateHashCode();
    }

    @Override
    public int getType() {
        return myType;
    }

    @Override
    @NonNull
    public String getArtifactAddress() {
        return myArtifactAddress;
    }

    @Override
    @NonNull
    public File getArtifact() {
        return myArtifactFile;
    }

    @Override
    @Nullable
    public String getVariant() {
        throw unsupportedMethodForJavaLibrary("getVariant");
    }

    @Nullable
    @Override
    public String getBuildId() {
        throw unsupportedMethodForJavaLibrary("getBuildId");
    }

    @Override
    @NonNull
    public String getProjectPath() {
        throw unsupportedMethodForJavaLibrary("getProjectPath");
    }

    @Override
    @NonNull
    public File getFolder() {
        throw unsupportedMethodForJavaLibrary("getFolder");
    }

    @Override
    @NonNull
    public String getManifest() {
        throw unsupportedMethodForJavaLibrary("getManifest");
    }

    @Override
    @NonNull
    public String getJarFile() {
        throw unsupportedMethodForJavaLibrary("getJarFile");
    }

    @Override
    @NonNull
    public String getCompileJarFile() {
        throw unsupportedMethodForJavaLibrary("getCompileJarFile");
    }

    @Override
    @NonNull
    public String getResFolder() {
        throw unsupportedMethodForJavaLibrary("getResFolder");
    }

    @Nullable
    @Override
    public File getResStaticLibrary() {
        throw unsupportedMethodForJavaLibrary("getResStaticLibrary");
    }

    @Override
    @NonNull
    public String getAssetsFolder() {
        throw unsupportedMethodForJavaLibrary("getAssetsFolder");
    }

    @Override
    @NonNull
    public Collection<String> getLocalJars() {
        throw unsupportedMethodForJavaLibrary("getLocalJars");
    }

    @Override
    @NonNull
    public String getJniFolder() {
        throw unsupportedMethodForJavaLibrary("getJniFolder");
    }

    @Override
    @NonNull
    public String getAidlFolder() {
        throw unsupportedMethodForJavaLibrary("getAidlFolder");
    }

    @Override
    @NonNull
    public String getRenderscriptFolder() {
        throw unsupportedMethodForJavaLibrary("getRenderscriptFolder");
    }

    @Override
    @NonNull
    public String getProguardRules() {
        throw unsupportedMethodForJavaLibrary("getProguardRules");
    }

    @Override
    @NonNull
    public String getLintJar() {
        throw unsupportedMethodForJavaLibrary("getLintJar");
    }

    @Override
    @NonNull
    public String getExternalAnnotations() {
        throw unsupportedMethodForJavaLibrary("getExternalAnnotations");
    }

    @Override
    @NonNull
    public String getPublicResources() {
        throw unsupportedMethodForJavaLibrary("getPublicResources");
    }

    @Override
    @NonNull
    public String getSymbolFile() {
        throw unsupportedMethodForJavaLibrary("getSymbolFile");
    }

    @NonNull
    private static UnsupportedOperationException unsupportedMethodForJavaLibrary(
            @NonNull String methodName) {
        return new UnsupportedOperationException(
                methodName + "() cannot be called when getType() returns LIBRARY_JAVA");
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof IdeJavaLibrary)) {
            return false;
        }
        IdeJavaLibrary that = (IdeJavaLibrary) o;
        return myType == that.myType
                && Objects.equals(myArtifactAddress, that.myArtifactAddress)
                && Objects.equals(myArtifactFile, that.myArtifactFile);
    }

    @Override
    public int hashCode() {
        return myHashCode;
    }

    private int calculateHashCode() {
        return Objects.hash(myType, myArtifactAddress, myArtifactFile);
    }

    @Override
    public String toString() {
        return "IdeJavaLibrary{"
                + "myType="
                + myType
                + ", myArtifactAddress='"
                + myArtifactAddress
                + '\''
                + ", myArtifactFile="
                + myArtifactFile
                + '}';
    }
}

File: com.android.tools.build/4.2.0-alpha10/com.android.tools/sdk-common/27.2.0-alpha10/com/android/ide/common/gradle/model/impl/IdeJavaLibrary.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model.impl

import com.android.ide.common.gradle.model.IdeLibrary
import com.google.common.annotations.VisibleForTesting
import java.io.File
import java.io.Serializable

/**
 * The implementation of IdeLibrary for Java libraries.
 **/
data class IdeJavaLibrary(
  val core: IdeJavaLibraryCore,
  override val isProvided: Boolean
) : IdeLibrary by core, Serializable {
  @VisibleForTesting
  constructor(
    artifactAddress: String,
    artifact: File,
    isProvided: Boolean
  ) : this(IdeJavaLibraryCore(artifactAddress, artifact), isProvided)
}

data class IdeJavaLibraryCore(
  override val artifactAddress: String,
  override val artifact: File
) : IdeLibrary, Serializable {
  // Used for serialization by the IDE.
  internal constructor() : this(
    artifactAddress = "",
    artifact = File("")
  )

  override val type: IdeLibrary.LibraryType
    get() = IdeLibrary.LibraryType.LIBRARY_JAVA

  override val variant: String?
    get() = throw unsupportedMethodForJavaLibrary("getVariant")

  override val buildId: String?
    get() = throw unsupportedMethodForJavaLibrary("getBuildId")

  override val projectPath: String
    get() = throw unsupportedMethodForJavaLibrary("getProjectPath")

  override val folder: File?
    get() = null

  override val manifest: String
    get() = throw unsupportedMethodForJavaLibrary("getManifest")

  override val jarFile: String
    get() = throw unsupportedMethodForJavaLibrary("getJarFile")

  override val compileJarFile: String
    get() = throw unsupportedMethodForJavaLibrary("getCompileJarFile")

  override val resFolder: String
    get() = throw unsupportedMethodForJavaLibrary("getResFolder")

  override val resStaticLibrary: File?
    get() = throw unsupportedMethodForJavaLibrary("getResStaticLibrary")

  override val assetsFolder: String
    get() = throw unsupportedMethodForJavaLibrary("getAssetsFolder")

  override val localJars: Collection<String>
    get() = throw unsupportedMethodForJavaLibrary("getLocalJars")

  override val jniFolder: String
    get() = throw unsupportedMethodForJavaLibrary("getJniFolder")

  override val aidlFolder: String
    get() = throw unsupportedMethodForJavaLibrary("getAidlFolder")

  override val renderscriptFolder: String
    get() = throw unsupportedMethodForJavaLibrary("getRenderscriptFolder")

  override val proguardRules: String
    get() = throw unsupportedMethodForJavaLibrary("getProguardRules")

  override val lintJar: String?
    get() = null

  override val externalAnnotations: String
    get() = throw unsupportedMethodForJavaLibrary("getExternalAnnotations")

  override val publicResources: String
    get() = throw unsupportedMethodForJavaLibrary("getPublicResources")

  override val symbolFile: String
    get() = throw unsupportedMethodForJavaLibrary("getSymbolFile")

  override val isProvided: Nothing
    get() = error("abstract")
}

private fun unsupportedMethodForJavaLibrary(methodName: String): UnsupportedOperationException {
  return UnsupportedOperationException("$methodName() cannot be called when getType() returns LIBRARY_JAVA")
}
